gem/optimise.py

```diff
@@ -493,3 +590,101 @@ def aggressive_unroll(expression):
+@count_flop_node.register(ListTensor)
+@count_flop_node.register(FlexiblyIndexed)
+@count_flop_node.register(LogicalNot)
+@count_flop_node.register(LogicalAnd)
+@count_flop_node.register(LogicalOr)
+@count_flop_node.register(Conditional)
+def count_flop_node_zero(node):
+    return 0
+
+
+@count_flop_node.register(Power)
+@count_flop_node.register(Comparison)
+@count_flop_node.register(Sum)
+@count_flop_node.register(Product)
+@count_flop_node.register(Division)
+@count_flop_node.register(MathFunction)
+def count_flop_node_single(node):
+    return numpy.prod([idx.extent for idx in node.free_indices])
+
+
+@count_flop_node.register(IndexSum)
+def count_flop_node_index_sum(node):
+    return numpy.prod([idx.extent for idx in node.multiindex + node.free_indices])
+
+
+def count_flop(node):
+    """
+    Count the total floating point operations required to compute a gem node.
+    This function assumes that all subnodes that occur more than once induce a
+    temporary, and are therefore only computed once.
+    """
+    return sum(map(count_flop_node, traversal([node])))
+
+
+def fast_sum_factorise(sum_indices, factors):
+    return sum_factorise(*delta_elimination(list(reversed(sum_indices)), factors))
+
+
+def optimise(node, quadrature_multiindex, argument_multiindices):
+    from gem.refactorise import ATOMIC, COMPOUND, OTHER, collect_monomials
+    argument_indices = tuple([i for indices in argument_multiindices for i in indices])
+
+    def classify(argument_indices, expression):
+        if isinstance(expression, Conditional):
+            return ATOMIC
+        n = len(argument_indices.intersection(expression.free_indices))
+        if n == 0:
+            return OTHER
+        elif n == 1:
+            if isinstance(expression, Indexed):
+                return ATOMIC
+            else:
+                return COMPOUND
+        else:
+            return COMPOUND
+    classifier = partial(classify, set(argument_indices))
+
+    monomial_sum, = collect_monomials([node], classifier)
+    monomial_sum.argument_indices = argument_indices
+    monomial_sum = monomial_sum.optimise()
+
+    return monomial_sum.to_expression()
+
+
+def optimise_expressions(expressions, quadrature_multiindices, argument_multiindices):
+    """
+    perform loop optimisations on gem DAGs
+    :param expressions: list of gem DAGs
+    :param quadrature_multiindices: quadrature multiindices, tuple of tuples
+    :param argument_multiindices: argument multiindices, tuple of tuples
+    :return: list of optimised gem DAGs
+    """
+    if propagate_failure(expressions):
+        return expressions
+    return [optimise(node, quadrature_multiindices, argument_multiindices) for node in expressions]
+
+
+def propagate_failure(expressions):
+    """
+    Check if any gem nodes is Failure. In that case there is no need for subsequent optimisation.
+    """
+    for n in traversal(expressions):
+        if isinstance(n, Failure):
+            return True
+    return False
```

Comment:
These three functions should live in `coffee_mode.py`, they are not generic tools, but rather specific optimisation passes.